src/transformers/models/character_bert/tokenization_character_bert.py

```diff
@@ -0,0 +1,979 @@
+
+        If `token_ids_1` is `None`, this method only returns the first portion of the mask (0s).
+
+        Args:
+            token_ids_0 (`List[int]`):
+                List of IDs.
+            token_ids_1 (`List[int]`, *optional*):
+                Optional second list of IDs for sequence pairs.
+
+        Returns:
+            `List[int]`: List of [token type IDs](../glossary#token-type-ids) according to the given sequence(s).
+        """
+        sep = [self.sep_token_id]
+        cls = [self.cls_token_id]
+        if token_ids_1 is None:
+            return len(cls + token_ids_0 + sep) * [0]
+        return len(cls + token_ids_0 + sep) * [0] + len(token_ids_1 + sep) * [1]
+
+    def save_vocabulary(self, save_directory: str, filename_prefix: Optional[str] = None) -> Tuple[str]:
+        logger.warning("CharacterBERT does not have a token vocabulary. Skipping saving `vocab.txt`.")
+        return ()
+
+    def save_mlm_vocabulary(self, save_directory: str, filename_prefix: Optional[str] = None) -> Tuple[str]:
+        # NOTE: CharacterBERT has no token vocabulary, this is just to allow
+        # saving tokenizer configuration via CharacterBertTokenizer.save_pretrained
+        index = 0
+        if os.path.isdir(save_directory):
+            vocab_file = os.path.join(
+                save_directory,
+                (filename_prefix + "-" if filename_prefix else "") + VOCAB_FILES_NAMES["mlm_vocab_file"],
+            )
+        else:
+            vocab_file = (filename_prefix + "-" if filename_prefix else "") + save_directory
+        with open(vocab_file, "w", encoding="utf-8") as writer:
+            for token, token_index in sorted(self.mlm_vocab.items(), key=lambda kv: kv[1]):
+                if index != token_index:
+                    logger.warning(
+                        f"Saving MLM vocabulary to {vocab_file}: vocabulary indices are not consecutive."
+                        " Please check that the vocabulary is not corrupted!"
+                    )
+                    index = token_index
+                writer.write(token + "\n")
+                index += 1
+        return (vocab_file,)
+
+    def _save_pretrained(
+        self,
+        save_directory: Union[str, os.PathLike],
+        file_names: Tuple[str],
+        legacy_format: Optional[bool] = None,
+        filename_prefix: Optional[str] = None,
+    ):
+        file_names = super()._save_pretrained(
+            save_directory=save_directory,
+            file_names=file_names,
+            legacy_format=legacy_format,
+            filename_prefix=filename_prefix,
+        )
+
+        mlm_vocab_files = self.save_mlm_vocabulary(save_directory, filename_prefix=filename_prefix)
+        return file_names + mlm_vocab_files
+
+    # NOTE: this override is a workaround to make the attention mask computation
+    # rely on the tokens axis (number of tokens in the sequence) rather than the
+    # characters axis (number of characters per word).
+    def pad(
```

Comment:
not completely sure I understand this either, if the sequence is correctly tokenized in charaters, the sequence lenght should be the number of characters, and the mask can just check for special tokens which will not be individual characters

Reply:
I think it's perhaps a bit clearer now, but basically this is because I have list(list(int)) instead of list(int) and therefore I need to adapt things to look at the appropriate axis in the input tensors which are 3d instead of 2d.